src/metabase/upload.clj

```diff
@@ -483,3 +489,133 @@
                               :num-rows    (count (rest rows))}))]
           (snowplow/track-event! ::snowplow/csv-upload-failed api/*current-user-id* fail-stats))
         (throw e)))))
+
+;;; +--------------------------
+;;; |  append to uploaded table
+;;; +--------------------------
+
+(defn- base-type->upload-type
+  "Returns the most specific upload type for the given base type."
+  [base-type]
+  (condp #(isa? %2 %1) base-type
+    :type/Float                  ::float
+    :type/BigInteger             ::int
+    :type/Integer                ::int
+    :type/Boolean                ::boolean
+    :type/DateTimeWithTZ         ::offset-datetime
+    :type/DateTime               ::datetime
+    :type/Date                   ::date
+    :type/Text                   ::text))
+
+(defn- check-schema
+  "Returns a map with two keys:
+    - `parse-rows`, a function that takes a lazy-seq of rows, and returns a lazy-seq of rows. It parses the
+      string values in each row and removes columns that have the same normalized name as the generated columns.
+    - `table-col-names`, a seq of the column names in the table, in the same order as the columns in the CSV file.
+
+  Throws an exception if:
+    - the CSV file contains duplicate column names
+    - the schema of the CSV file does not match the schema of the table"
+  [table header]
+  ;; Assumes table-cols are unique when normalized
+  (let [table-cols           (t2/select :model/Field :table_id (:id table))
```

Comment:
```suggestion
  (let [table-fields           (t2/select :model/Field :table_id (:id table))
```

I rarely see we use the word "col" for Field here, I think we should stick with `fields` for 2 reasons:
- cols should be reserved for column in csv file
- table-cols can also be understand as "table collection"